### ISSUE TITLE
Linker errors while trying to compile Tesseract project

### ISSUE THREAD

**Issue by saint1729** [tesseract-ocr/tesseract#686]
Hi,

I have put my effort by making changes to necessary project files. I have used the solution file present at the following location:

https://github.com/peirick/VS2015_Tesseract/

Unfortunately, that is based on Tesseract - 3.05. I took the latest 4.0 version of source code to build in visual studio. I'm getting following linker errors only for the newly introduced code part in 4.0 i.e., lstm and simddetect:

Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1120	8 unresolved externals	tesseract	D:\tesseract\VS2015_Tesseract\x64\Debug\tesseract.exe	1
Error	LNK2001	unresolved external symbol "private: static bool SIMDDetect::avx_available_" (?avx_available_@SIMDDetect@@0_NA)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\tesseractmain.obj 1
Error	LNK2001	unresolved external symbol "private: static bool SIMDDetect::sse_available_" (?sse_available_@SIMDDetect@@0_NA)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\tesseractmain.obj 1
Error	LNK2019	unresolved external symbol "public: void __cdecl tesseract::Tesseract::TrainLineRecognizer(class STRING const &,class STRING const &,class BLOCK_LIST *)" (?TrainLineRecognizer@Tesseract@tesseract@@QEAAXAEBVSTRING@@0PEAVBLOCK_LIST@@@Z) referenced in function "public: int __cdecl tesseract::TessBaseAPI::Recognize(class ETEXT_DESC *)" (?Recognize@TessBaseAPI@tesseract@@QEAAHPEAVETEXT_DESC@@@Z)	tesseract D:\tesseract\VS2015_Tesseract\tesseract\baseapi.obj	1
Error	LNK2019	unresolved external symbol "public: void __cdecl tesseract::Tesseract::LSTMRecognizeWord(class BLOCK const &,class ROW *,class WERD_RES *,class tesseract::PointerVector *)" (?LSTMRecognizeWord@Tesseract@tesseract@@QEAAXAEBVBLOCK@@PEAVROW@@PEAVWERD_RES@@PEAV?$PointerVector@VWERD_RES@@@2@@Z) referenced in function "public: void __cdecl tesseract::Tesseract::classify_word_pass1(struct tesseract::WordData const &,class WERD_RES * *,class tesseract::PointerVector *)" (?classify_word_pass1@Tesseract@tesseract@@QEAAXAEBUWordData@2@PEAPEAVWERD_RES@@PEAV?$PointerVector@VWERD_RES@@@2@@Z)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\control.obj	1
Error	LNK2019	unresolved external symbol "public: __cdecl tesseract::LSTMRecognizer::LSTMRecognizer(void)" (??0LSTMRecognizer@tesseract@@QEAA@XZ) referenced in function "public: bool __cdecl tesseract::Tesseract::init_tesseract_lang_data(char const *,char const *,char const *,enum tesseract::OcrEngineMode,char * *,int,class GenericVector const *,class GenericVector const *,bool)" (?init_tesseract_lang_data@Tesseract@tesseract@@QEAA_NPEBD00W4OcrEngineMode@2@PEAPEADHPEBV?$GenericVector@VSTRING@@@@3_N@Z)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\tessedit.obj 1
Error	LNK2019	unresolved external symbol "public: bool __cdecl tesseract::LSTMRecognizer::DeSerialize(bool,class tesseract::TFile *)" (?DeSerialize@LSTMRecognizer@tesseract@@QEAA_N_NPEAVTFile@2@@Z) referenced in function "public: bool __cdecl tesseract::Tesseract::init_tesseract_lang_data(char const *,char const *,char const *,enum tesseract::OcrEngineMode,char * *,int,class GenericVector const *,class GenericVector const *,bool)" (?init_tesseract_lang_data@Tesseract@tesseract@@QEAA_NPEBD00W4OcrEngineMode@2@PEAPEADHPEBV?$GenericVector@VSTRING@@@@3_N@Z)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\tessedit.obj 1
Error	LNK2019	unresolved external symbol "public: bool __cdecl tesseract::LSTMRecognizer::LoadDictionary(char const *,char const *)" (?LoadDictionary@LSTMRecognizer@tesseract@@QEAA_NPEBD0@Z) referenced in function "public: bool __cdecl tesseract::Tesseract::init_tesseract_lang_data(char const *,char const *,char const *,enum tesseract::OcrEngineMode,char * *,int,class GenericVector const *,class GenericVector const *,bool)" (?init_tesseract_lang_data@Tesseract@tesseract@@QEAA_NPEBD00W4OcrEngineMode@2@PEAPEADHPEBV?$GenericVector@VSTRING@@@@3_N@Z)	tesseract	D:\tesseract\VS2015_Tesseract\tesseract\tessedit.obj 1
Error	LNK2019	unresolved external symbol "public: __cdecl tesseract::LSTMRecognizer::~LSTMRecognizer(void)" (??1LSTMRecognizer@tesseract@@QEAA@XZ) referenced in function "public: void * __cdecl tesseract::LSTMRecognizer::`scalar deleting destructor'(unsigned int)" (??_GLSTMRecognizer@tesseract@@QEAAPEAXI@Z)	tesseract D:\tesseract\VS2015_Tesseract\tesseract\tesseractclass.obj	1

Can someone kindly assist me in resolving them?

**Comment by Shreeshrii** [comment]
Try installing with cppan as described in

https://github.com/tesseract-ocr/tesseract/wiki/Compiling

- excuse the brevity, sent from mobile

**Comment by egorpugin** [comment]
Yes. Please, download cppan client and run `cppan --build pvt.cppan.demo.google.tesseract-master`.

**Comment by saint1729** [comment]
I ran the above command but still no luck. 2 weeks gone but still no progress on build. Can you kindly describe the step by step procedure so that I can have a successful build. Thanks in advance.

**Comment by egorpugin** [comment]
First, write what output you get from those.

**Comment by christophered** [comment]
Can you provide a video of installing Tesseract 4.0x along with it's step-by-step description, on Ubuntu 16.xx from scratch. Uploading it to Youtube so that all people interested in the matter can follow from scratch.

**Comment by saint1729** [comment]
Hi Egor,

Thanks for your response. Forgetting what happened in the past, I now ran the following steps for building Tesseract from source:

git clone https://github.com/tesseract-ocr/tesseract tesseract
cd tesseract
cppan
mkdir build && cd build
cmake ..

Build is now successful.

But, when I'm trying to run tesseractmain.cpp program in Debug mode, I get the following error:

![1](https://cloud.githubusercontent.com/assets/1130987/22403458/d9d32832-e63d-11e6-8411-5782ec20e184.png)

![2](https://cloud.githubusercontent.com/assets/1130987/22403461/e287262c-e63d-11e6-9eff-104cf87a5207.png)

Can you kindly assist me in getting rid of this error and run the main program?

**Comment by egorpugin** [comment]
1. build them - Yes (Also turn on that check box.)
2. Select tesseractmain project in Solution Explorer - in the context menu set it as default project.

**Comment by saint1729** [comment]
Perfect. Setting tesseractmain as the default project worked. Thanks a ton for the help. Closing the thread.

**Comment by BrianZhu01** [comment]
I also meet the same problem, but I can't resolve it with your answer. can you give me some suggestion?

**Comment by saint1729** [comment]
Hi @BrianZhu01 , can you let us know what all steps you've tried from beginning?

**Comment by BrianZhu01** [comment]
thanks, I download tesseract, but I do not how deal with cppan? ,I had download cppan.exe .what can I do?

**Comment by BrianZhu01** [comment]
I do steps as you list, but have the follow problem, and I just put the cppan.exe in cmake path? what I have wrong?@saint1729
![0x943d51f0_0x4da8_0x43f6_0xa3_0x22b8_0x2278_0x225d_0x2286_0x224a_0x222d_0x2284](https://cloud.githubusercontent.com/assets/15243563/22917419/b3241f50-f2bf-11e6-846d-98476318c3ee.png)

**Comment by saint1729** [comment]
Add **cppan** to you system variable **PATH**.

Go to directory where you extracted tesseract source code
**cd tesseract**


Run the following commands after that:

**cppan
mkdir build && cd build
cmake ..**

**Comment by saint1729** [comment]
Can you list what all files are present in your tesseract directory by running **dir** command and upload the screenshot here?

**Comment by BrianZhu01** [comment]
thanks a lot,  If I run the cmd using administrator, I reslove the problem. when I run cppan , it can download the need files. thanks for your help,

**Comment by BrianZhu01** [comment]
@saint1729 ,sorry, I also need you help. run cppan doesn't work now! yesterday, I just download some files through cppan. but I meet some errors when run "cmake".
![000000](https://cloud.githusercontent.com/assets/15243563/22970142/fe2009f0-f3ab-11e6-8b74-81f628f701b7.png)
![00000001](https://cloud.githubusercontent.com/assets/15243563/22970145/fefbee02-f3ab-11e6-8f8c-f2080af15767.png)
![0000002](https://cloud.githubusercontent.com/assets/15243563/22970143/fe30a5e4-f3ab-11e6-800b-1c863b6c7296.png)

**Comment by BrianZhu01** [comment]
@saint1729 
![000000](https://cloud.githubusercontent.com/assets/15243563/22970246/649f3002-f3ac-11e6-9312-c82e8cb5e325.png)

**Comment by egorpugin** [comment]
This is because of early error of cppan.
Try to open `cppan.yml` in some editor and save with unix line endings (LF).
Also try to upgrade cppan `cppan --self-upgrade`.

**Comment by BrianZhu01** [comment]
@egorpugin ， I opencv cppan.yml use txt. so the format have changed. what editor can i use to recover it?thanks!

**Comment by egorpugin** [comment]
notepad++

**Comment by BrianZhu01** [comment]
@saint1729 @egorpugin ，I had open cppan.yml use notepad++, but here also have same error!
![0000000000000000000](https://cloud.githubusercontent.com/assets/15243563/23002929/5cf771da-f428-11e6-9bec-0374c86f9147.png)

**Comment by SarahMestiri** [comment]
What if I have cppan that stops when launching it? I can't install Tesseract using cppan. I really need help! Thank you. 
Windows 10 64bit
opencv installed correctly.
Visual studio community 2017

**Comment by monu143** [comment]
Yeah, **cppan** crashes when I tried to install in windows. So I used this way instead
Run the following commands in CMD/Powershell after downloading and extracting the repo.

Repo: https://github.com/Microsoft/vcpkg.git

Commands:
```.\bootstrap-vcpkg.bat```
Next run:
```.\vcpkg integrate install```
Finally run this command:
```.\vcpkg install tesseract:x64-windows --head```

The last command would take a lot of time and download a lot of packages. Once completed you'd have tesseract installed in the same folder as the repo extract.

Path to include files: 
```C:\VisualStudioLib\vcpkg\installed\x64-windows\include\tesseract```
Path to Lib files:
```C:\VisualStudioLib\vcpkg\installed\x64-windows\lib```
